rx_signal_processing/filtering.cpp

```diff
@@ -0,0 +1,280 @@
+ */
+void Filtering::mix_first_stage_to_bandpass(const std::vector<double> &rx_freqs,
+                                              double initial_rx_sample_rate) {
+    for (uint32_t i=0; i<rx_freqs.size(); i++) { // TODO(keith): comment the protobuf with what rx freqs are. Offset or center.
+      auto sampling_freq = 2 * M_PI * rx_freqs[i]/initial_rx_sample_rate; //radians per sample
+
+      //TODO(keith): verify that this is okay.
+      for(int j=0;j < first_stage_lowpass_taps.size(); j++) {
+        auto radians = fmod(sampling_freq * j,2 * M_PI);
+        auto amplitude = first_stage_lowpass_taps[j].real();
+        auto I = amplitude * cos(radians);
+        auto Q = amplitude * sin(radians);
+        first_stage_bandpass_taps_h.push_back(std::complex<float>(I,Q));
+      }
+    }
+}
+
+/**
+ * @brief      Writes out a set of filter taps to file in case they need to be tested.
+ *
+ * @param[in]  filter_taps  A reference to a vector of filter taps.
+ * @param[in]  name         A output file name.
+ */
+void Filtering::save_filter_to_file(const std::vector<std::complex<float>> &filter_taps, std::string name) {
+  std::ofstream filter;
+  filter.exceptions(std::ofstream::failbit | std::ofstream::badbit);
+  try {
+    filter.open(name);
+  } catch (std::system_error &e) {
+    //TODO(keith): handle error
+  }
+
+  if (filter_taps.size() == 0) {
+    //TODO(keith): handle error
+  }
+  for (auto &i : filter_taps){
+    filter << i << std::endl;
+  }
+  filter.close();
+}
+
+/**
+ * @brief      Gets the number of first stage taps.
+ *
+ * @return     The number of first stage taps.
+ */
+uint32_t Filtering::get_num_first_stage_taps() {
+  return num_first_stage_taps;
+}
+
+/**
+ * @brief      Gets the number of second stage taps.
+ *
+ * @return     The number of second stage taps.
+ */
+uint32_t Filtering::get_num_second_stage_taps() {
+  return num_second_stage_taps;
+}
+
+/**
+ * @brief      Gets the number of third stage taps.
+ *
+ * @return     The number of third stage taps.
+ */
+uint32_t Filtering::get_num_third_stage_taps() {
+  return num_third_stage_taps;
+}
+
+/**
+ * @brief      Gets the vector of the first stage lowpass filter taps.
+ *
+ * @return     The first stage lowpass taps.
+ */
+std::vector<std::complex<float>> Filtering::get_first_stage_lowpass_taps() {
+  return first_stage_lowpass_taps;
+}
+
+/**
+ * @brief      Gets the vector of the second stage lowpass filter taps.
+ *
+ * @return     The second stage lowpass taps.
+ */
+std::vector<std::complex<float>> Filtering::get_second_stage_lowpass_taps() {
+  return second_stage_lowpass_taps;
+}
+
+/**
+ * @brief      Gets the vector of the third stage lowpass taps.
+ *
+ * @return     The third stage lowpass taps.
+ */
+std::vector<std::complex<float>> Filtering::get_third_stage_lowpass_taps() {
+  return third_stage_lowpass_taps;
+}
+
+/**
+ * @brief      Gets the vector containing bandpass filter taps for each RX frequency.
```

Comment:
Since this is a vector storing all the taps for all frequencies, can you write a sentence like: "For example, with 3 rx frequencies and 32 taps, this vector will be 96 taps long". I don't know how the remez 'taps + 1' comes into play here but maybe it's important as well.